### ISSUE TITLE
Import .proto files from Maven dependencies

### ISSUE THREAD

**Issue by mwillema** [os72/protoc-jar-maven-plugin#19]
Is there a way to import .proto files from Maven dependencies? This would be a practical way to share .proto files between Maven projects.

All I could find is a [similar question on Stack Overflow](http://stackoverflow.com/questions/13165333/importing-protocol-buffer-definitions-between-maven-projects) but it is about the [maven-protoc-plugin](https://github.com/dtrott/maven-protoc-plugin), not this one. This other plugin looks for .proto files in Maven dependencies and put them into a temporary directory. This directory is then set as an import path to the protoc invocation.

It would be great to have a similar feature here :-)

**Comment by pandrez** [comment]
It would be really useful indeed!

**Comment by os72** [comment]
Ack

**Comment by vjkoskela** [comment]
For example, the [com.google.protobuf:protobuf-java](http://search.maven.org/#artifactdetails%7Ccom.google.protobuf%7Cprotobuf-java%7C3.1.0%7Cbundle) jar contains the [wrapper.proto](https://github.com/google/protobuf/blob/master/src/google/protobuf/wrappers.proto) definition.

**Comment by os72** [comment]
The standard types from Google are bundled and supported via `<includeStdTypes>true</includeStdTypes>`

**Comment by wooky** [comment]
I'm trying to get this to work. I have a protobuf file with `import "google/protobuf/any.proto"` and includeStdTypes is true, but it complains that it can't find that file.

**Comment by os72** [comment]
How is your plugin configured, and what does the maven log say?

It's working for me:
```
[INFO]     Processing (java): grpc_hello.proto
protoc-jar: executing: [C:\cygwin64\tmp\protocjar2763937346440857941\bin\protoc.exe, -IC:\cygwin64\tmp\protocjar2763937346440857941\include, -IC:\home\work\github\test-misc\src\test\resources, --java_out=C:\home\work\github\test-misc\target\generated-test-sources, C:\home\work\github\test-misc\src\test\resources\grpc_hello.proto]
grpc_hello.proto: warning: Import google/protobuf/any.proto but not used.
```

**Comment by wooky** [comment]
Running this from IntelliJ:
```
[INFO] Scanning for projects...
[INFO]                                                                         
[INFO] ------------------------------------------------------------------------
[INFO] Building xtremettt-common 1.0-SNAPSHOT
[INFO] ------------------------------------------------------------------------
[INFO] 
[INFO] --- protoc-jar-maven-plugin:3.3.0.1:run (default-cli) @ xtremettt-common ---
[INFO] Protoc version: 330
protoc-jar: protoc version: 330, detected platform: windows 10/amd64
protoc-jar: executing: [C:\Users\me\AppData\Local\Temp\protocjar5431910111010375489\bin\protoc.exe, --version]
libprotoc 3.3.0
[INFO] Protoc command: C:\Users\me\AppData\Local\Temp\protocjar5431910111010375489\bin\protoc.exe
[INFO] Input directories:
[INFO]     C:\Users\me\IdeaProjects\xtremettt\src\main\protobuf
[INFO] Output targets:
[INFO]     java: C:\Users\me\IdeaProjects\xtremettt\target\generated-sources (add: main, clean: false, plugin: null, outputOptions: null)
[INFO]     Processing (java): CommonRequest.proto
protoc-jar: executing: [C:\Users\me\AppData\Local\Temp\protocjar5431910111010375489\bin\protoc.exe, -IC:\Users\me\IdeaProjects\xtremettt\src\main\protobuf, --java_out=C:\Users\me\IdeaProjects\xtremettt\target\generated-sources, C:\Users\me\IdeaProjects\xtremettt\src\main\protobuf\CommonRequest.proto]
google/protobuf/any.proto: File not found.
CommonRequest.proto: Import "google/protobuf/any.proto" was not found or had errors.
CommonRequest.proto:7:5: "google.protobuf.Any" is not defined.
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 0.604 s
[INFO] Finished at: 2017-06-27T20:26:22-06:00
[INFO] Final Memory: 8M/155M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal com.github.os72:protoc-jar-maven-plugin:3.3.0.1:run (default-cli) on project xtremettt-common: protoc-jar failed for C:\Users\me\IdeaProjects\xtremettt\src\main\protobuf\CommonRequest.proto. Exit code 1 -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

Process finished with exit code 1

```

**Comment by os72** [comment]
Somehow I don't think `includeStdTypes` is set to true. The cmd line should show an additional include like:
`-IC:\Users\me\AppData\Local\Temp\protocjar5431910111010375489\include`

**Comment by wooky** [comment]
So for some reason, if I run the protoc-jar:run target from IntelliJ, it doesn't see the flag, but if I run the compile lifecycle, it does see it.

**Comment by docker8436** [comment]
Maven build log shows "google/protobuf/any.proto: File not found." 
Protoc-jar log didn't pickup the temp folder expected. (see below)
protoc-jar: executing: [/var/folders/fb/czdpvlf102z8tf91tlkqj8lw0000gn/T/protocjar5770121856020262014/bin/protoc.exe, -I/Users/.../src/main/protobuf/com/lf/iotcore/domaindriven/logicaldevicedata, --java_out=/Users/.../target/generated-sources, /Users/.../src/main/protobuf/com/lf/iotcore/domaindriven/logicaldevicedata/LocationUpdate.proto]

The POM is 
   <plugin>
                <groupId>com.github.os72</groupId>
                <artifactId>protoc-jar-maven-plugin</artifactId>
                <version>3.4.0</version>
                <executions>
                    <execution>
                        <phase>generate-sources</phase>
                        <goals>
                            <goal>run</goal>
                        </goals>
                        <configuration>
                            <protocVersion>3.4.0</protocVersion> 
                            <includeStdTypes>true</includeStdTypes>
                        </configuration>
                    </execution>
                </executions>
            </plugin>

And I do have protobuf-java and rotobuf-java-util 3.4.0 installed.

**Comment by os72** [comment]
Are you sure it ran with `includeStdTypes` true? It should show an additional include path like `/var/folders/fb/czdpvlf102z8tf91tlkqj8lw0000gn/T/protocjar5770121856020262014/include`

**Comment by docker8436** [comment]
just don't know anything else I can do other than set the ```xml<includeStdTypes>true</includeStdTypes>``` in the pom file. Everything else works, except for the import stuff.

**Comment by os72** [comment]
Are you using IntelliJ? Is this on mac? Does it work with mvn on cmd line?

**Comment by docker8436** [comment]
IntelliJ. and Mac. But I only run Mvn form cmd line.

**Comment by docker8436** [comment]
my bad. the way I invoked mvn is wrong. If I used "mvn generate-sources", everything works. but if I used "mvn protoc-jar:run", it will have this problem. sorry for the confusion.

**Comment by RishuK** [comment]
Hi, 
My import statements from some custom proto files that are present in a jar are not getting imported. How can that be achieved?

<plugin>
            <groupId>com.github.os72</groupId>
            <artifactId>protoc-jar-maven-plugin</artifactId>
            <version>${protoc-maven-plugin.version}</version>
            <executions>
                <execution>
                    <phase>generate-sources</phase>
                    <goals>
                        <goal>run</goal>
                    </goals>
                    <configuration>
                        <protocVersion>${protobuf.version}</protocVersion>
                        <includeStdTypes>true</includeStdTypes>
                        <includeDirectories>
                            <include>src/main/resources</include>
                        </includeDirectories>
                        <inputDirectories>
                            <include>src/main/resources</include>
                        </inputDirectories>
                    </configuration>
                </execution>
            </executions>
        </plugin>

**Comment by os72** [comment]
Anyone interested to help with this implementation?

**Comment by RishuK** [comment]
Would be great if someone can help

**Comment by os72** [comment]
Thanks @jvanbaa13 for helping implement this!

**Comment by os72** [comment]
Released protoc-jar-maven-plugin 3.6.0

Supports new option "includeMavenTypes", thanks @jvanbaa13 

`<includeMavenTypes>direct</includeMavenTypes>`
`<includeMavenTypes>transitive</includeMavenTypes>`

**Comment by os72** [comment]
Released protoc-jar-maven-plugin 3.6.0.1
